OsmAnd-java/src/main/java/net/osmand/router/RouteColorize.java

```diff
@@ -0,0 +1,416 @@
+     */
+    private double[] findDerivativeArguments(double[] distances, double[] elevations, int index, double slopeRange) {
+        double[] result = new double[4];
+        double minDist = distances[index] - slopeRange / 2;
+        double maxDist = distances[index] + slopeRange / 2;
+        result[0] = Double.NaN;
+        result[1] = Double.NaN;
+        result[2] = minDist;
+        result[3] = maxDist;
+        int closestMaxIndex = -1;
+        int closestMinIndex = -1;
+        for (int i = index; i < distances.length; i++) {
+            if (distances[i] == maxDist) {
+                result[1] = elevations[i];
+                break;
+            }
+            if (distances[i] > maxDist) {
+                closestMaxIndex = i;
+                break;
+            }
+        }
+        for (int i = index; i >= 0; i--) {
+            if (distances[i] == minDist) {
+                result[0] = elevations[i];
+                break;
+            }
+            if (distances[i] < minDist) {
+                closestMinIndex = i;
+                break;
+            }
+        }
+        if (closestMaxIndex > 0) {
+            double diff = distances[closestMaxIndex] - distances[closestMaxIndex - 1];
+            double coef = (maxDist - distances[closestMaxIndex - 1]) / diff;
+            if (coef > 1 || coef < 0) {
+                //System.out.println("Coefficient fo max must be 0..1 , coef=" + coef);
+            }
+            result[1] = (1 - coef) * elevations[closestMaxIndex - 1] + coef * elevations[closestMaxIndex];
+        }
+        if (closestMinIndex >= 0) {
+            double diff = distances[closestMinIndex + 1] - distances[closestMinIndex];
+            double coef = (minDist - distances[closestMinIndex]) / diff;
+            if (coef > 1 || coef < 0) {
+                //System.out.println("Coefficient for min must be 0..1 , coef=" + coef);
+            }
+            result[0] = (1 - coef) * elevations[closestMinIndex] + coef * elevations[closestMinIndex + 1];
+        }
+        if (Double.isNaN(result[0]) || Double.isNaN(result[1])) {
+            //System.out.println("Elevations wasn't calculated");
+        }
+        return result;
+    }
+
+    private void calculateMinMaxValue() {
+        if (values.length == 0)
+            return;
+        minValue = maxValue = Double.NaN;
+        for (double value : values) {
+            if ((Double.isNaN(maxValue) || Double.isNaN(minValue)) && !Double.isNaN(value))
+                maxValue = minValue = value;
+            if (minValue > value)
+                minValue = value;
+            if (maxValue < value)
+                maxValue = value;
+        }
+    }
+
+    private double[] listToArray(List<Double> doubleList) {
+        double[] result = new double[doubleList.size()];
+        for (int i = 0; i < doubleList.size(); i++) {
+            result[i] = doubleList.get(i);
+        }
+        return result;
+    }
+
+    public class Data {
```

Comment:
Data is probably bad name for class, 
1) it should be static inner class
2) name suggestion: RouteColorizationSegment / Point